// ngx-highlight-js/__ivy_ngcc__/index.js
import { NgModule } from '@angular/core';
import { HighlightJsDirective } from './components/highlight-js.directive';
import * as ɵngcc0 from '@angular/core';
export { HighlightJsDirective } from './components/highlight-js.directive';
var HighlightJsModule = (function () {
    function HighlightJsModule() {
    }
    /** @nocollapse */
    HighlightJsModule.ctorParameters = function () { return []; };
HighlightJsModule.ɵfac = function HighlightJsModule_Factory(t) { return new (t || HighlightJsModule)(); };
HighlightJsModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: HighlightJsModule });
HighlightJsModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(HighlightJsModule, [{
        type: NgModule,
        args: [{
                declarations: [HighlightJsDirective],
                exports: [HighlightJsDirective]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(HighlightJsModule, { declarations: function () { return [HighlightJsDirective]; }, exports: function () { return [HighlightJsDirective]; } }); })();
    return HighlightJsModule;
}());
export { HighlightJsModule };
